### PR TITLE
Elaborate supertrait obligations when deducing closure signatures

### LINKED ISSUE #23012: Type inference error: the type of this value must be known in this context
Repro:

``` rust
pub trait Receive<T: Send, E: Send> : FnOnce(Result<T, E>) + Send {
    fn receive(self, res: Result<T, E>);
}

impl<T: Send, E: Send, F: FnOnce(Result<T, E>) + Send> Receive<T, E> for F {
    fn receive(self, res: Result<T, E>) {
        self(res)
    }
}

pub trait Async<T, E> {
    fn receive<F: Receive<T, E>>(self, f: F);
}

impl<T, E> Async<T, E> for Result<T, E> {
    fn receive<F: Receive<T, E>>(self, f: F) {
        f(self)
    }
}

fn res() -> Result<u32, ()> {
    Ok(123)
}

pub fn main() {
    res().receive(|res| println!("{}", res.unwrap()))
}
```

Output:

```
$ rustc fn.rs
fn.rs:26:40: 26:52 error: the type of this value must be known in this context
fn.rs:26     res().receive(|res| println!("{}", res.unwrap()))
                                                ^~~~~~~~~~~~
note: in expansion of format_args!
<std macros>:2:43: 2:76 note: expansion site
<std macros>:1:1: 2:78 note: in expansion of println!
fn.rs:26:25: 26:54 note: expansion site
error: aborting due to previous error
```

### PR DESCRIPTION
**Pull request by compiler-errors**
We elaborate the supertrait obligations of any registered predicates for a closure to see if we can infer a closure signature.

This is not as general of a fix as it *could* be, since we just elaborate supertrait bounds instead of doing a theoretical walk of _all_ registered predicates that might cause us to deduce `Fn` trait information for a closure infer var. I don't even know how to come up with an example that fails here but would work with a more general system.

Fixes #23012
Also fixes the existing compile failure in #57611

r? @ghost for now until I do a perf run
cc @nikomatsakis since you commented on https://github.com/rust-lang/rust/issues/23012#issuecomment-1152397290

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? types

I'd like an initial review here just for the code changes, then I'll need to open an FCP to see if the team is okay with the generalization of inference.

**Comment by compiler-errors** [comment]
Gonna re-roll.

r? types

**Comment by spastorino** [comment]
re-rolling :joy:

r? types

**Comment by spastorino** [comment]
@compiler-errors if you re-rolled it when it was assigned to @jackh726, I guess it's because we don't want to assign to Jack?.

**Comment by compiler-errors** [comment]
😆 

r? types

**Comment by jackh726** [comment]
agh sorry for no review here :(

I can try to give some comments this weekend

File: compiler/rustc_hir_analysis/src/check/closure.rs [hunk 1]
```diff
@@ -226,33 +226,44 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                }
+                ty::PredicateKind::Trait(data) => Some(data.def_id()),
+                _ => None,
+            };
+            if let Some(closure_kind) =
+                trait_def_id.and_then(|def_id| self.tcx.fn_trait_kind_from_lang_item(def_id))
+            {
+                expected_kind = Some(
+                    expected_kind
+                        .map_or_else(|| closure_kind, |current| cmp::min(current, closure_kind)),
```
Comment (lcnr) [inline compiler/rustc_hir_analysis/src/check/closure.rs:270]:
```suggestion
                        .map_or(closure_kind, |current| cmp::min(current, closure_kind)),
```

**Review by lcnr — COMMENTED** [review]
r=me on the impl, do you have a test impacted by e508304e8291c78bcae8784f4eaec5f81913d809? otherwise, please merge that commit with the previous one.

**Comment by compiler-errors** [comment]
> do you have a test impacted by https://github.com/rust-lang/rust/commit/e508304e8291c78bcae8784f4eaec5f81913d809?

Yup, the test called `multiple-fn-bounds.rs` has the Fn trait bounds in its error message flipped.

https://github.com/rust-lang/rust/blob/693485373b971d1574388b9da785b66c19dc86d2/src/test/ui/closures/multiple-fn-bounds.stderr#L14

**Pull request merged**